File: solutions/005-longest-palindromic-substring/solution.py
# 思路：中心扩展。每个位置作为回文中心，向两侧扩展；分奇长（i,i）和偶长（i,i+1）两种情况
class Solution:
    def longest_palindrome(self, s: str) -> str:
        if len(s) == 0:
            return ""
        start, max_len = 0, 1

        def expand(left: int, right: int) -> None:
            nonlocal start, max_len
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            l = right - left - 1
            if l > max_len:
                max_len = l
                start = left + 1

        for i in range(len(s)):
            expand(i, i)
            expand(i, i + 1)
        return s[start : start + max_len]


if __name__ == "__main__":
    tests = [("babad", "bab"), ("cbbd", "bb")]
    sol = Solution()
    for s, expected in tests:
        got = sol.longest_palindrome(s)
        status = "PASS" if got == expected else "FAIL"
        print(f"{status} | s={s!r} => {got!r} (expected {expected!r})")

File: solutions/005-longest-palindromic-substring/solution.go
package main

import "fmt"

// 思路：中心扩展。每个位置作为回文中心，向两侧扩展；分奇长（i,i）和偶长（i,i+1）两种情况
func longestPalindrome(s string) string {
	if len(s) == 0 {
		return ""
	}
	start, maxLen := 0, 1

	expand := func(left, right int) {
		for left >= 0 && right < len(s) && s[left] == s[right] {
			left--
			right++
		}
		// 循环结束时多扩出去一位，合法区间是 [left+1, right-1]
		if l := right - left - 1; l > maxLen {
			maxLen = l
			start = left + 1
		}
	}

	for i := range s {
		expand(i, i)   // 奇数长度，如 "aba"
		expand(i, i+1) // 偶数长度，如 "bb"
	}
	return s[start : start+maxLen]
}

func main() {
	got := longestPalindrome("babad")
	status := "PASS"
	if got != "bab" && got != "aba" {
		status = "FAIL"
	}
	fmt.Printf("%s | s=%q => %q\n", status, "babad", got)
}
